### PR TITLE
Makefile, contrib: Add script to create kind cluster

### PR DESCRIPTION
**Pull request by christarazi**
We have the equivalent for minikube.

Usage:
```
$ ./contrib/scripts/kind.sh
$ # or
$ make kind
```

### CODE REVIEW

**Comment by Rolinh** [comment]
>     * Find a way to parameterize the number of workers (not sure how
>       necessary this is)

I use kind a lot for Hubble Relay development and, in this context, being able to test with 4+ nodes is super useful (I've created kind clusters up to 50 nodes).

However, I also found out that I don't need a local registry as I can just `kind load docker-image cilium/hubble-relay:latest` (and any other relevant image) followed by `kubectl delete pods ...`. This works because the pull policy is `IfNotPresent` and if there are multiple images with the same tag, the most recent one is picked.

**Comment by sayboras** [comment]
I love kind a lot as well. Have been using it as part of local development and testing.

Related on number of node, here is the configuration we used in smoke test with one tweak to make sure that the control plan is available for scheduling. https://github.com/cilium/cilium/blob/master/.github/kind-config.yaml#L6

**Comment by joestringer** [comment]
FWIW, I feel it'd be kinda nice to have some convenience script for this in the tree, occasionally I use the steps from the conformance tests + `.github/kind-config.yaml` to deploy a kind config but I always have to look it up again whenever I try it out.

**Comment by christarazi** [comment]
Yeah, I think it's worth dusting off this PR and getting it out even though it doesn't meet all the TODOs marked in the PR description. I often switch to this branch locally, create my kind cluster, then switch back to my dev branch :sweat_smile:. 

/cc @sayboras, he had a similar PR out and might be worth figuring out which one to move forward and whether to consolidate.

**Comment by sayboras** [comment]
@joestringer @christarazi 👋 oh this pending one I got is my local script to quickly run conformance test. There are few comments which I need to address, but didn't have time recently. I will plan sometimes to clean it up, anyway, happy to discuss how to move forward in either case 😃

**Comment by sayboras** [comment]
Forgot to update this one. Recently, the idea is to have a good scripting to test locally, but recently I switch to cilium-cli, which is great tool. I think this one is no longer required, we can promote cilium-cli further as well.

**Comment by christarazi** [comment]
It might still be helpful to have an "official" script to launch a dev kind cluster. I still plan to get to this at some point :sweat_smile:

**Comment by sayboras** [comment]
> It might still be helpful to have an "official" script to launch a dev kind cluster. I still plan to get to this at some point 😅

Oh sorry, just realized that I closed the wrong PR :(, there is another one created by me 🤦

File: contrib/scripts/kind.sh [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+for node in $(kind get nodes); do
+  kubectl annotate node "${node}" "kind.x-k8s.io/registry=localhost:${reg_port}";
+done
+
+set +e
+kubectl taint nodes --all node-role.kubernetes.io/master-
+set -e
+
+echo
+echo "Images are pushed into the kind registry like so:"
```
Comment (aditighag) [inline contrib/scripts/kind.sh:105]: I usually run `kind load...` to push images. With the kind registry set up, `make dev-docker-image` will push images to that registry instead?

**Review by aditighag — APPROVED** [review]
👍 Are you planning to add more steps to install cilium to the same script?

File: contrib/scripts/kind.sh [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+kind_cmd="kind create cluster"
+
+if [[ -n "${CLUSTER_NAME}" ]]; then
+  kind_cmd+=" --name ${CLUSTER_NAME}"
+fi
+if [[ -n "${IMAGE}" ]]; then
+  kind_cmd+=" --image ${IMAGE}"
+fi
+
```
Comment (aditighag) [inline contrib/scripts/kind.sh:62]: Might also need to set context - `kubectl cluster-info --context kind-kind`

File: contrib/scripts/kind.sh (hunk 1 above)
Comment (christarazi) [reply contrib/scripts/kind.sh:105]: I've had trouble when using `kind load ...` so I've been using this script for..a year now I guess with the registry. `make dev-docker-image` will tag/push images as long as you set the `DOCKER_REGISTRY` variable to `localhost:5000`. It won't do it automatically

File: contrib/scripts/kind.sh (hunk 2 above)
Comment (christarazi) [reply contrib/scripts/kind.sh:62]: I think the `kind` command when it creates the cluster, it automatically sets the context

File: contrib/scripts/kind.sh (hunk 1 above)
Comment (christarazi) [reply contrib/scripts/kind.sh:105]: In terms of adding more steps, no I don't think that's really necessary. I've opted to keep the script simple: just create a kind cluster. Especially now that we have the Cilium CLI, it's so easy to install Cilium into a cluster

File: contrib/scripts/kind.sh (hunk 2 above)
Comment (aditighag) [reply contrib/scripts/kind.sh:62]: Yeah, that has been my observation lately as well.

**Review by michi-covalent — APPROVED** [review]

**Pull request merged**